vnext/ReactUWP/Modules/AsyncStorageModuleWin32.cpp

```diff
@@ -0,0 +1,191 @@
+    if (db) {
+      ExecImpl(db, u8"BEGIN TRANSACTION", nullptr, nullptr);
+    }
+  }
+  Sqlite3Transaction(const Sqlite3Transaction &) = delete;
+  Sqlite3Transaction(Sqlite3Transaction &&other) : m_db{other.m_db} {
+    other.m_db = nullptr;
+  }
+  Sqlite3Transaction &operator=(const Sqlite3Transaction &) = delete;
+  Sqlite3Transaction &operator=(Sqlite3Transaction &&rhs) {
+    std::swap(m_db, rhs.m_db);
+  }
+  ~Sqlite3Transaction() {
+    ExecImpl(m_db, u8"COMMIT", nullptr, nullptr);
+  }
+};
+
+} // namespace
+
+AsyncStorageModuleWin32::AsyncStorageModuleWin32(PCSTR dbPath) {
+  checkSQLiteResult(
+      sqlite3_open_v2(dbPath, &m_db, SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_FULLMUTEX, nullptr));
+
+  int userVersion = 0;
+  auto userVersionCallback = [&](int cCol, char **rgszColText, char **rgszColName) {
+    userVersion = atoi(rgszColText[0]);
+    return SQLITE_OK;
+  };
+
+  Exec(u8"PRAGMA user_version", userVersionCallback);
+
+  if (userVersion == 0) {
+    Exec(u8"CREATE TABLE IF NOT EXISTS AsyncLocalStorage(key TEXT, value TEXT); PRAGMA user_version=1");
+  }
+}
+
+AsyncStorageModuleWin32::~AsyncStorageModuleWin32() {
+  checkSQLiteResult(sqlite3_close(m_db));
+}
+
+std::string AsyncStorageModuleWin32::getName() {
+  return "AsyncLocalStorage";
+}
+
+std::map<std::string, dynamic> AsyncStorageModuleWin32::getConstants() {
+  return {};
+}
+
+std::vector<facebook::xplat::module::CxxModule::Method> AsyncStorageModuleWin32::getMethods() {
+  return {Method("multiGet", this, &AsyncStorageModuleWin32::multiGet),
+          Method("multiSet", this, &AsyncStorageModuleWin32::multiSet),
+          Method("multiRemove", this, &AsyncStorageModuleWin32::multiRemove),
+          // The 'multiMerge' method is currently not implemented. We assume that
+          // the ReactNative framework translates its absence into the AsyncStorage
+          // JS object (see Libraries\Storage\AsyncStorage.js around line 485) which
+          // in turn allows JS user code to test for support of merge methods.
+          Method("clear", this, &AsyncStorageModuleWin32::clear),
+          Method("getAllKeys", this, &AsyncStorageModuleWin32::getAllKeys)};
+}
+
+void AsyncStorageModuleWin32::multiGet(folly::dynamic params, Callback callback) {
+  folly::dynamic result = folly::dynamic::array;
+  auto &&args = params[0];
+  if (args.size() != 0) {
+    Sqlite3Transaction transaction(m_db);
+    auto pStmt = PrepareStatement(u8"SELECT key, value FROM AsyncLocalStorage WHERE key = ?");
+    for (auto &&arg : args) {
+      BindString(pStmt, 1, arg.getString());
+      checkSQLiteResult(sqlite3_step(pStmt.get()));
+      auto key = reinterpret_cast<const char *>(sqlite3_column_text(pStmt.get(), 0));
+      if (!key)
+        checkSQLiteResult(sqlite3_errcode(m_db));
+      auto value = reinterpret_cast<const char *>(sqlite3_column_text(pStmt.get(), 1));
+      if (!value)
+        checkSQLiteResult(sqlite3_errcode(m_db));
```

Comment:
Could we propagate these errors to the callback?

Reply:
All exceptions from these methods are going to result in std::terminate. We need proper error handling and make them noexcept.

---
In reply to: [379338049](https://github.com/microsoft/react-native-windows/pull/4102#discussion_r379338049) [](ancestors = 379338049)